### DISCUSSION TITLE
`exportPathMap` deprecation and replacement for filename mapping functionality across directories

### DISCUSSION THREAD

**Discussion by bjester** [vercel/next.js#51108] [Help] [answered]
### Summary

I saw in the documentation that `exportPathMap` is deprecated in favor of using `getStaticPaths`. I understand how to use that instead, but it seems it's missing one capability that `exportPathMap` had-- the ability to do a simple mapping of source => export filename that involves directory structure differences.

We built a SSG site that used `exportPathMap` to write the exported files to a names that are more like codes than nice url names (see code snippet). In this manner, it also allowed us to organize the source pages in subdirectories, as we expect to have many groupings of these pages. With `getStaticPaths` I believe I could organize the pages like `/<group name>/[slug].js`, but how can we have the export write those to the root directory?

I could use rewrites in the webserver, but that could get unwieldy and I'm not sure how nicely that plays with next.js. We don't need to worry about cross-page routing either.

### Additional information

```js
exportPathMap: async function (defaultPathMap, { dev, dir, outDir, distDir, buildId }) {
    return {
      '/abcd': { page: '/<group name>/<nice filename>.js' },
      '/efgh': { page: '/<group name>/<another nice filename>.js' },
    };
  }
```

**Comment by devjiwonchoi** [comment, marked as answer]
Since you've been "hard-coding" the paths to `exportPathMap` for your routes, you can use the `middleware` to do the same.

Say your `group name` is `abcd`, and follow the convention of Next.js Page Router Dynamic Route.
The route path will be `/abcd/[slug]`, and this is not we want.

**BUT** we can use the middleware to `rewrite` as you have stated, corresponding to the path `/abcd`.

Example:

```js
// middleware.ts

import { NextResponse } from 'next/server'
import type { NextRequest } from 'next/server'
 
export function middleware(request: NextRequest) {
  if (request.nextUrl.pathname === '/abcd') {
    return NextResponse.rewrite(new URL('/abcd/abcd', request.url))
  }

  if (request.nextUrl.pathname === '/efgh') {
    return NextResponse.rewrite(new URL('/efgh/efgh', request.url))
  }
}
```

By doing this, you can have separate files for your preserved dynamic route using SSG. 
This is possible because you've been hard-coding your route preserved to specific files as `abcd` to `nice filename.js` and `efgh` to `another nice filename.js`.

**Reply by bjester** [reply]
Thanks for your reply. I see your example code imports `'next/server'`, so to verify this works for static site generation (SSG)? Server side rendering isn't an option for us

**Reply by devjiwonchoi** [reply]
Yes, but to clarify, I just want to know if your routes are predetermined like `/abcd`, `/efgh`, etc. Or, is it just a [slug] and the paths can vary?

**Reply by bjester** [reply]
The `/abcd`, `/efgh`, etc. routes are predetermined yes.

**Reply by devjiwonchoi** [reply]
Then my answer would work fine with SSG since the actual route that is rewriting to is `/abcd/[slug].js` with the regular SSG you are using.

**Reply by bjester** [reply]
This does seem to work, thanks!

**Comment by nestorbritez** [comment]
[Rewrites](https://nextjs.org/docs/pages/api-reference/next-config-js/rewrites) allow you to map an incoming request path to a different destination path.
```
// next.config.js

module.exports = {
  async rewrites() {
    return [
      {
        source: '/about',
        destination: '/',
      },
    ]
  },
}
```
